estransport/connection.go

```diff
@@ -0,0 +1,300 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package estransport
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/url"
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	defaultResurrectTimeoutInitial      = 60 * time.Second
+	defaultResurrectTimeoutFactorCutoff = 5
+)
+
+// ConnectionPool defines the interface for the connection pool.
+//
+type ConnectionPool interface {
+	Next() (*Connection, error)
+	Remove(*Connection) error
+}
+
+// Connection represents a connection to a node.
+//
+type Connection struct {
+	sync.Mutex
+
+	URL       *url.URL  `json:"url"`
+	Dead      bool      `json:"dead"`
```

Comment:
Not sure about that, to be honest. I can't see a way for outside code to access an instance of `Connection` — they are hidden behind unexported fields of `estransport.Client`. I like that these fields (and in future, other fields, coming from cluster state) are exported, for documentation purposes, and also, when the package allows replacing connection pool implementation, which is planned. I've removed the JSON tags in c7cf16b, though.

Reply:
As soon as it's exported, it's part of the API. Maybe unexport `Connection` (rename to `connection`) until needed, and at that point it can be reassessed? If nothing else, documentation about concurrency expectations would be good.